grpc/internal/gateway/gateway.go

```diff
@@ -0,0 +1,89 @@
+package gateway
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/grpc-ecosystem/grpc-gateway/runtime"
+	"github.com/keratin/authn-server/api"
+	"github.com/keratin/authn-server/config"
+	authnpb "github.com/keratin/authn-server/grpc"
+)
+
+type GatewayResponseMiddleware func(ctx context.Context, response http.ResponseWriter, m proto.Message) error
+
+// StatusCodeMutator changes the HTTP Repsonse status code to the desired mapping. The default mapping
+// by gRPC-Gateway doesn't have some of the desired responses (e.g. 201), which is why this function is needed.
+func StatusCodeMutator(ctx context.Context, response http.ResponseWriter, m proto.Message) error {
+	log.Printf("%s: message type: %T", time.Now(), m)
+	switch m.(type) {
+	case *authnpb.SignupResponseEnvelope, *authnpb.LoginResponseEnvelope, *authnpb.RefreshSessionResponseEnvelope, *authnpb.SubmitPasswordlessLoginResponseEnvelope, *authnpb.ChangePasswordResponseEnvelope:
+		response.WriteHeader(http.StatusCreated)
+	}
```

Comment:
Perhaps I should fallback to `runtime.HTTPStatusFromCode(code codes.Code)` here?